tidy3d/plugins/microwave/array_factor.py

```diff
@@ -761,28 +878,494 @@ def array_factor(
-        af_y = np.sum(
-            amp_y
-            * np.exp(-1j * np.arange(self.array_size[1])[:, None, None] * psi_y[np.newaxis, :]),
-            axis=0,
-        )
-        af_z = np.sum(
-            amp_z
-            * np.exp(-1j * np.arange(self.array_size[2])[:, None, None] * psi_z[np.newaxis, :]),
-            axis=0,
+        # Compute array factor based on the defined taper
+        if self.taper is None or isinstance(self.taper, RectangularTaper):
+            return self._rect_taper_array_factor(exp_x, exp_y, exp_z)
+        else:
+            return self._general_taper_array_factor(exp_x, exp_y, exp_z)
+
+
+class AbstractWindow(Tidy3dBaseModel, ABC):
+    """This class provides interface for window selection."""
+
+    @abstractmethod
+    def get_weights(self, N: int) -> ArrayLike:
+        """Interface function for computing window weights at N points."""
+
+
+class HammingWindow(AbstractWindow):
+    """Standard Hamming window for tapering or spectral shaping."""
+
+    def get_weights(self, N: int) -> ArrayLike:
+        """
+        Generate a 1D Hamming window of length N.
+
+        Parameters
+        ----------
+        N : int
+            Number of points in the window.
+
+        Returns
+        -------
+        ArrayLike
+            1D array of Hamming window weights.
+        """
+        return hamming(N)
+
+
+class BlackmanWindow(AbstractWindow):
+    """Standard Blackman window for tapering or spectral shaping."""
+
+    def get_weights(self, N: int) -> ArrayLike:
+        """
+        Generate a 1D Blackman window of length N.
+
+        Parameters
+        ----------
+        N : int
+            Number of points in the window.
+
+        Returns
+        -------
+        ArrayLike
+            1D array of Blackman window weights.
+        """
+        return blackman(N)
+
+
+class BlackmanHarrisWindow(AbstractWindow):
+    """Standard Blackman-Harris window for tapering or spectral shaping."""
+
+    def get_weights(self, N: int) -> ArrayLike:
+        """
+        Generate a 1D Blackman-Harris window of length N.
+
+        Parameters
+        ----------
+        N : int
+            Number of points in the window.
+
+        Returns
+        -------
+        ArrayLike
+            1D array of Blackman-Harris window weights.
+        """
+        return blackmanharris(N)
+
+
+class HannWindow(AbstractWindow):
+    """Hann window with configurable sidelobe suppression and sidelobe count."""
+
+    def get_weights(self, N: int) -> ArrayLike:
+        """
+        Generate a 1D Hann window of length N.
+
+        Parameters
+        ----------
+        N : int
+            Number of points in the window.
+
+        Returns
+        -------
+        ArrayLike
+            1D array of Hann window weights.
+        """
+        return hann(N)
+
+
+class ChebWindow(AbstractWindow):
+    """Standard Chebyshev window for tapering with configurable sidelobe attenuation."""
+
+    at: pd.PositiveFloat = pd.Field(
```

Comment:
more informative to name the field "attenuation", unless "at" is widely used in the field.